applications/pelion_client/src/modules/power_manager.c

```diff
@@ -0,0 +1,254 @@
+
+#define MODULE power_manager
+#include <caf/events/module_state_event.h>
+
+#include <logging/log_ctrl.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_PELION_CLIENT_POWER_MANAGER_LOG_LEVEL);
+
+#include "keep_alive_event.h"
+
+
+#define POWER_DOWN_ERROR_TIMEOUT \
+	K_SECONDS(CONFIG_PELION_CLIENT_POWER_MANAGER_ERROR_TIMEOUT)
+#define POWER_DOWN_TIMEOUT_MS \
+	(CONFIG_PELION_CLIENT_POWER_MANAGER_TIMEOUT * MSEC_PER_SEC)
+#define POWER_DOWN_CHECK_MS	 1000
+
+enum power_state {
+	POWER_STATE_IDLE,
+	POWER_STATE_SUSPENDING,
+	POWER_STATE_SUSPENDED,
+	POWER_STATE_OFF,
+	POWER_STATE_WAKEUP,
+	POWER_STATE_ERROR,
+	POWER_STATE_ERROR_SUSPENDED,
+	POWER_STATE_ERROR_OFF
+};
+
+static enum power_state power_state = POWER_STATE_IDLE;
+static struct k_delayed_work power_down_trigger;
+static struct k_delayed_work error_trigger;
+static atomic_t power_down_count;
+
+
+static bool system_off_check(void)
+{
+	return !IS_ENABLED(CONFIG_PELION_CLIENT_POWER_MANAGER_STAY_ON) &&
+	       (power_state == POWER_STATE_SUSPENDED);
+}
+
+static void system_off(void)
+{
+	if (power_state == POWER_STATE_ERROR_SUSPENDED) {
+		power_state = POWER_STATE_ERROR_OFF;
+	} else {
+		power_state = POWER_STATE_OFF;
+	}
+
+	LOG_WRN("System turned off");
+	LOG_PANIC();
+
+	pm_power_state_force((struct pm_state_info){PM_STATE_SOFT_OFF, 0, 0});
+}
+
+static void power_down(struct k_work *work)
+{
+	atomic_val_t cnt = atomic_add(&power_down_count, POWER_DOWN_CHECK_MS);
+
+	if (cnt >= POWER_DOWN_TIMEOUT_MS) {
+		if (system_off_check()) {
+			system_off();
+		} else if (power_state != POWER_STATE_SUSPENDED) {
+			LOG_INF("System power down");
+
+			struct power_down_event *event = new_power_down_event();
+
+			event->error = false;
+			power_state = POWER_STATE_SUSPENDING;
+			EVENT_SUBMIT(event);
+		} else {
+			/* Stay suspended. */
+		}
+	} else {
+		k_delayed_work_submit(&power_down_trigger,
+				      K_MSEC(POWER_DOWN_CHECK_MS));
+	}
+}
+
+static void power_down_counter_reset(void)
+{
+	switch (power_state) {
+	case POWER_STATE_IDLE:
+		atomic_set(&power_down_count, 0);
+		break;
+	case POWER_STATE_SUSPENDING:
+	case POWER_STATE_SUSPENDED:
+	case POWER_STATE_OFF:
+	case POWER_STATE_ERROR:
+		/* No action */
+		break;
+	default:
+		__ASSERT_NO_MSG(false);
+		break;
+	}
+}
+
+struct pm_state_info pm_policy_next_state(int32_t ticks)
+{
+	return (struct pm_state_info){PM_STATE_ACTIVE, 0, 0};
+}
+
+bool pm_policy_low_power_devices(enum pm_state pm_state)
+{
+	return pm_is_sleep_state(pm_state);
+}
+
+static void error(struct k_work *work)
+{
+	struct power_down_event *event = new_power_down_event();
+
+	/* Turning off all the modules (including leds). */
+	event->error = false;
+	EVENT_SUBMIT(event);
+	power_state = POWER_STATE_ERROR_SUSPENDED;
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_keep_alive_event(eh)) {
+		power_down_counter_reset();
+
+		return false;
+	}
+
+	if (is_power_down_event(eh)) {
+		switch (power_state) {
```

Comment:
Can You put handling of this event to separate function? `handle_power_down_event` ? Similarly for others?

Reply:
Can do but this is migration from repo to repo, not a code cleanup. I don't want to make too much changes here.
Once we get it in, have CI execute against the changes, we can make beautifications.